doc/read-the-docs-site/howtos/asdata.rst

```diff
@@ -0,0 +1,89 @@
+2. If it is important to check that the entire structure is well-formed, then it is better to convert it, since the conversion will check for well-formedness.
+3. If correctness is important, then it can be better to convert it, since the compiler can help you more if you are using Plutus Tx's datatypes.
+
+Generally, which approach is better is an empirical question and may vary in different cases.
+Often the cost of ``fromBuiltinData`` can be very significant and so it is desirable to use a ``Data`` representation, but the third point above can be quite painful.
+
+The ``plutus-tx`` library provides a tool to make it possible to use a raw ``Data`` representation while still getting the same amount of help from the compiler.
+The ``asData`` function takes the definition of a data type and replaces it with an equivalent definition whose representation uses ``Data`` directly.
+
+For example, if we wanted to use it on the types from the :ref:`auction example<simple_example>`, we would put the datatype declarations inside a Template Haskell quote and call ``asData`` on it.
+
+.. literalinclude:: ../tutorials/AuctionValidator.hs
+   :start-after: BLOCK9
+   :end-before: BLOCK10
+
+This is normal Template Haskell that just generates new Haskell source, so you can see the code that it generates with `{-# OPTIONS_GHC -ddump-splices #-}`, but it will look something like this:
+
+.. code-block::
+
+        PlutusTx.asData
+        [d| data Bid'
+                = Bid' {bBidder' :: PubKeyHash, bAmount' :: Lovelace}
+                deriving newtype (Eq, Ord, ToBuitinData, FromBuiltinData, UnsafeFromBuiltinData)
+            data AuctionRedeemer' = NewBid' Bid | Payout'
+                deriving newtype (Eq, Ord, ToBuitinData, FromBuiltinData, UnsafeFromBuiltinData) |]
+
+        ======>
+
+        newtype Bid' = Bid'2 BuiltinData
+        deriving newtype (Eq, Ord, PlutusTx.ToData, FromData, UnsafeFromData)
+
+        {-# COMPLETE Bid' #-}
+        pattern Bid' :: PubKeyHash -> Lovelace -> Bid'
+        pattern Bid' ...
+
+        newtype AuctionRedeemer' = AuctionRedeemer'2 BuiltinData
+        deriving newtype (Eq, Ord, PlutusTx.ToData, FromData, UnsafeFromData)
+
+        {-# COMPLETE NewBid', Payout' #-}
+        pattern NewBid' :: Bid -> AuctionRedeemer'
+        pattern NewBid' ...
+        pattern Payout' :: AuctionRedeemer'
+        pattern Payout' ...
+
+That is:
+
+- It creates a newtype wrapper around ``BuiltinData``
+- It creates pattern synonyms corresponding to each of the constructors you wrote
+
+This lets you write code "as if" you were using the original declaration that you wrote, while in fact the pattern synonyms are handling conversion to/from ``Data`` for you.
+But any values of this type actually are represented with ``Data``.
+That means that when we newtype-derive the instances for converting to and from ``Data`` we get the instances for ``BuiltinData`` - which are free!
+
+Caveats
```

Comment:
There's another important caveat: it is advisable to access fields with pattern matching, since it is probably faster than field accessors. Record patterns can be used if there are too many fields.